nautobot/utilities/views.py

```diff
@@ -151,3 +186,1062 @@ def get_return_url(self, request, obj=None):
+                standard_fields = [field for field in form.fields if field not in custom_fields + ["pk"]]
+                nullified_fields = request.POST.getlist("_nullify")
+
+                try:
+
+                    with transaction.atomic():
+
+                        updated_objects = []
+                        for obj in self.queryset.filter(pk__in=form.cleaned_data["pk"]):
+
+                            obj = self.alter_obj_for_bulk_edit(obj, request, [], kwargs)
+
+                            # Update standard fields. If a field is listed in _nullify, delete its value.
+                            for name in standard_fields:
+
+                                try:
+                                    model_field = model._meta.get_field(name)
+                                except FieldDoesNotExist:
+                                    # This form field is used to modify a field rather than set its value directly
+                                    model_field = None
+
+                                # Handle nullification
+                                if name in form.nullable_fields and name in nullified_fields:
+                                    if isinstance(model_field, ManyToManyField):
+                                        getattr(obj, name).set([])
+                                    else:
+                                        setattr(obj, name, None if model_field.null else "")
+
+                                # ManyToManyFields
+                                elif isinstance(model_field, ManyToManyField):
+                                    if form.cleaned_data[name]:
+                                        getattr(obj, name).set(form.cleaned_data[name])
+                                # Normal fields
+                                elif form.cleaned_data[name] not in (None, ""):
+                                    setattr(obj, name, form.cleaned_data[name])
+
+                            # Update custom fields
+                            for name in custom_fields:
+                                if name in form.nullable_fields and name in nullified_fields:
+                                    obj.cf[name] = None
+                                elif form.cleaned_data.get(name) not in (None, ""):
+                                    obj.cf[name] = form.cleaned_data[name]
+
+                            obj.full_clean()
+                            obj.save()
+                            updated_objects.append(obj)
+                            logger.debug(f"Saved {obj} (PK: {obj.pk})")
+
+                            # Add/remove tags
+                            if form.cleaned_data.get("add_tags", None):
+                                obj.tags.add(*form.cleaned_data["add_tags"])
+                            if form.cleaned_data.get("remove_tags", None):
+                                obj.tags.remove(*form.cleaned_data["remove_tags"])
```

Comment:
We'll need to be careful throughout this file to keep the functionality here in sync with the corresponding code in `core.views.generic` - for example this is a block of code I'm going to need to modify substantially as part of #1318, and we'll need to replicate those diffs here.

Any ideas as to a pattern that would let us share code between the two code paths, instead of needing to do so much copy-pasting and keeping the two in sync manually?

Reply:
What might be nice here is to have a method on the object itself ferry bulk changes down to the mixins/features, although this might make the form/model separation a bit more grey.

```py
object.process_bulk_form_changes(form.cleaned_data)

#...

# Either as mixin or feature
class MyObject(TagsMixin, RelationshipMixin):

  #...

class BulkMixinBase():
  def process_bulk_form_changes(self, form_data):
    pass

class TagsMixin(BulkMixinBase):
  def process_bulk_form_changes(self, form_data):
    super().process_bulk_form_changes(form_data)

    if form_data.get("add_tags", None):
      self.tags.add(*form_data["add_tags"])
    #...

class RelationshipMixin(BulkMixinBase):
  def process_bulk_form_changes(self, form_data):
    super().process_bulk_form_changes(form_data)

    if form_data.get("add_relationships", None):
      self.relationships.add(*form_data["add_relationships"])
    #...

```